### ISSUE TITLE
PM2 Raspberry Jessie - Not spawning on boot

### ISSUE THREAD

**Issue by zettam** [Unitech/pm2#1654]
I've installed pm2 and nodejs v0.12.6 to my rpi2 with a brand new Jessie installation.

I ran the command `sudo pm2 startup` and it created my `pm2-init.sh` file.

Then I manually went into the file and changed the line:
`export PM2_HOME="/root/.pm2"`
to this:
`export PM2_HOME="/home/pi/.pm2"`

as "pi" is the default user in rpi2 jessie.

When I start my .js files with the command `pm2 start app.js` everything works however the boot functionality is still missing. Nothing is starting on boot.

Looking for a solution.

**Comment by zettam** [comment]
`pm2 list` is empty on boot

**Comment by zettam** [comment]
I believe for some reason the json file is not being filled when the processes are started.

**Comment by zettam** [comment]
update:
When I do `pm2.save`all of the processes are saved just fine. When I power off my rpi2 directly and power on (no `sudo reboot`) all saved processes are started.

However when I put in the commands `sudo poweroff` or `sudo reboot` to reboot my rpi2, process list is erased upon reboot. Nothing is resurrecting.

**Comment by zettam** [comment]
Any ideas?

**Comment by zettam** [comment]
Found the solution.
Reason of the problem: Jessie switched to systemd!

Solution (fairly simple!)

```sudo pm2 startup systemd -u pi```

this will make a systemd startup script for the user "pi". Magic.

**Comment by mastix** [comment]
You saved my life! :) :+1:

**Comment by scargill** [comment]
Saved mine too...

**Comment by tehfreak** [comment]
Thank you!

**Comment by zettam** [comment]
:)

**Comment by Unitech** [comment]
it is fixed via the new --hp option: https://github.com/Unitech/pm2/commit/419b748e58dc7874d6c5f25723032aa584c7c9ad#diff-4ce9aea92d36e9c98f1eae56904b4a3eR634

If anyone has still this issue please leave a comment

**Comment by bkawk** [comment]
pi@raspberrypi:~ $ sudo pm2 startup systemd -u pi
[PM2] Generating system init script in /etc/systemd/system/pm2.service
[PM2] Making script booting at startup...
[PM2] -systemd- Using the command:
      su pi -c "pm2 dump && pm2 kill" && su root -c "systemctl daemon-reload && systemctl enable pm2 && systemctl start pm2"
Command failed: /bin/sh -c su pi -c "pm2 dump && pm2 kill" && su root -c "systemctl daemon-reload && systemctl enable pm2 && systemctl start pm2"
events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: connect EACCES /home/pi/.pm2/rpc.sock
    at Object.exports._errnoException (util.js:874:11)
    at exports._exceptionWithHostPort (util.js:897:20)
    at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1063:14)

----- Are you sure you use the right platform command line option ? centos / redhat, amazon, ubuntu, gentoo, systemd or darwin?

**Comment by JFernandoGomez** [comment]
fixed with 

`sudo pm2 startup systemd -u pi`

 as @zettam  said!!

**Comment by zettam** [comment]
With the new raspberry 3 I tried the same, but it didn't seem to work. I came back to my own solution but lookslike it's nto worknig any more. Any ideas ?

**Comment by JFernandoGomez** [comment]
Worked for me on startup for pm2, but still had the "pm2 save" issue, process still gets killed on reboot. Ended implementing systemd solution =/

**Comment by zettam** [comment]
oh hmm...
without the systemd setting, it works for me with the following command:
pm2 startup -u pi
this is enough to create the startup script, and it starts the pm2 properly. However if I don't do manual save (or dump) it's not saving the final state of pm2 processes, and not spawning them properly. it's only spawning what I saved manually using the save or dump command.

**Comment by zettam** [comment]
`sudo pm2 startup systemd -u pi` does not seem to work any more. getting an error.

```
[PM2] Spawning PM2 daemon
[PM2] PM2 Successfully daemonized
[PM2] Generating system init script in /etc/systemd/system/pm2.service
[PM2] Making script booting at startup...
[PM2] -systemd- Using the command:
      su pi -c "pm2 dump && pm2 kill" && su root -c "systemctl daemon-reload && systemctl enable pm2 && systemctl start pm2"
Command failed: su pi -c "pm2 dump && pm2 kill" && su root -c "systemctl daemon-reload && systemctl enable pm2 && systemctl start pm2"
bash: pm2: command not found

----- Are you sure you use the right platform command line option ? centos / redhat, amazon, ubuntu, gentoo, systemd or darwin?

```

**Comment by zettam** [comment]
Setting dump file to immutable did not help.
Issue should be reopened.

**Comment by davinche** [comment]
Having an issue with this as well. Using latest version of Node/NPM 6.x. and latest version of pm2 via `sudo npm install -g pm2@latest`

**Comment by christroutner** [comment]
I tried just about every solution I could find on the internet. I finally got it by issuing this command:

`sudo env PATH=$PATH:/usr/local/bin pm2 startup systemd -u pi --hp /home/pi`

**Comment by sugarding** [comment]
@christroutner Thanks buddy! that worked like a charm

**Comment by Bonitis** [comment]
@christroutner that worked for me, but only if I power cycle the device. If I just run a reboot, I still lose the startup config and it says the dump file is empty. Any idea why there is a difference?

**Comment by natcl** [comment]
Same issue here with the latest Jessie

**Comment by dikkedimi** [comment]
Had this issue too on a RPi3, but ran
`sudo pm2 startup systemd -u pi`
and then
`sudo env PATH=$PATH:/usr/local/bin pm2 startup systemd -u pi --hp /home/pi`
and then
pm2 save

rebooted using `sudo shutdown -r now`

it now starts at boot, thanks guys!

**Comment by ngervasi** [comment]
I have been fighting with this for a couple of days and finally found how to fix (latest pm2 on jessie):
1 - `pm2 startup` and execute the suggested script with sudo
2 - `pm2 start yourApp.js`
3 - `pm2 save`
4 - Edit `/etc/systemd/system/pm2-pi.service` and comment out the following line:
`#Type=forking `

Now my app starts at boot.

**Comment by vmarchaud** [comment]
We will release a fix in the next days, checkout : https://github.com/Unitech/pm2/issues/2765

**Comment by kastnerstefan** [comment]
>>> Executing systemctl start pm2-pi
Job for pm2-pi.service failed. See 'systemctl status pm2-pi.service' and 'journalctl -xn' for details.
[ERROR] Exit code : 1
[PM2][ERROR] systemctl start pm2-pi failed, see error above.

Got this error by calling:
sudo env PATH=$PATH:/usr/local/bin /usr/local/lib/node_modules/pm2/bin/pm2 startup systemd -u pi --hp /home/pi

**Comment by vmarchaud** [comment]
Could you share the error logged in the journalctl ?

**Comment by kastnerstefan** [comment]
`pi@raspberrypi:~ $ journalctl -xn
No journal files were found.`

I get this message if i look for journalctl.

**Comment by vmarchaud** [comment]
Without errors i cant help you.

**Comment by kastnerstefan** [comment]
With `systemctl status pm2-pi.service` i get:
`pm2-pi.service - PM2 process manager`
`Loaded: loaded (/etc/systemd/system/pm2-pi.service; enabled)`
`Active: failed (Result: timeout) since Sa 2017-03-18 22:50:26 CET; 1min 10s ago`
`Docs: https://pm2.keymetrics.io/`
`Process: 1532 ExecStart=/usr/local/lib/node_modules/pm2/bin/pm2 resurrect (code=killed, signal=TERM)`

**Comment by jasonyork** [comment]
I had the same problem.  The issue was that the default startup timeout is set to 8 seconds and it was taking longer than that to start up.  Try this:
`sudo nano /etc/systemd/system/pm2-pi.service`
change
`TimeoutStartSec=20`
save the file
`sudo systemctl daemon-reload`
Reboot

**Comment by LaughingTree** [comment]
@jasonyork - I just completed a fresh install on a Rasp Pi 3. I had to delete the /nodes_module folder and run 
`npm install`
to fix the "Failed at electron post install script 'node install.js' " error. My download from the /MichMich/master didn't include the electron module for some reason. Completely vanilla install on a fresh pi 3 otherwise. 
Yet when I run your command
`sudo systemct1 daemon-reload`
I get 
`sudo: systemct1: command not found`

Any thoughts?
